aws-sns-topicpolicy/src/main/java/software/amazon/sns/topicpolicy/UpdateHandler.java

```diff
@@ -27,54 +23,70 @@ protected ProgressEvent<ResourceModel, CallbackContext> handleRequest(
-    /**
-     * Invocation of doDelete calls setTopicAttributes to delete topic-policy of a given topic ARN.
-     *
-     * @param proxy
-     *            {@link AmazonWebServicesClientProxy} to initiate proxy chain
-     * @param proxyClient
-     *            the aws service client {@link ProxyClient<SnsClient>} to make the call
-     * @param request
-     *            {@link ResourceHandlerRequest<ResourceModel>}
-     * @param progress
-     *            {@link ProgressEvent<ResourceModel, CallbackContext>} to place hold the current progress data
-     * @return {@link ProgressEvent<ResourceModel, CallbackContext>}
-     */
+        return ProgressEvent.progress(resourceModel, callbackContext)
+                .then(progress -> Update(proxy, proxyClient, request, progress, logger))
+                .then(progress -> ProgressEvent.defaultSuccessHandler(null));
+    }
 
-    private ProgressEvent<ResourceModel, CallbackContext> doDelete(
+    protected ProgressEvent<ResourceModel, CallbackContext> Update(
             final AmazonWebServicesClientProxy proxy,
             final ProxyClient<SnsClient> proxyClient,
             final ResourceHandlerRequest<ResourceModel> request,
-            final ProgressEvent<ResourceModel, CallbackContext> progress) {
+            ProgressEvent<ResourceModel, CallbackContext> progress,
+            final Logger logger) {
         final ResourceModel model = request.getDesiredResourceState();
         final ResourceModel previousState = request.getPreviousResourceState();
+        final String policy = getPolicyDocument(request);
         // new topics
         List<String> newTopics = model.getTopics();
         // previous topics
         List<String> previousTopics = previousState.getTopics();
         // extract the topics that needs to be deleted.
         previousTopics.removeAll(newTopics);
-        return handleDelete(proxy, proxyClient, request, progress, previousTopics, ACTION_DELETED, UPDATE_HANDLER, logger);
+        final CallbackContext callbackContext = progress.getCallbackContext();
+        for (final String topicArn : newTopics) {
+            final ProgressEvent<ResourceModel, CallbackContext> progressEvent = proxy
+                    .initiate("AWS-SNS-TopicPolicy::Update", proxyClient, model, callbackContext)
+                    .translateToServiceRequest((resourceModel) -> Translator.translateToRequest(topicArn, policy))
+                    .makeServiceCall((awsRequest, client) -> {
+                        SetTopicAttributesResponse response = proxyClient.injectCredentialsAndInvokeV2(awsRequest, client.client()::setTopicAttributes);
+                        logger.log(String.format("Resource Updated in StackId: %s", request.getStackId()));
+                        return response;
+                    })
+                    .handleError((awsRequest, exception, client, rModel, context) -> handleError(awsRequest, exception, client, rModel, context))
+                    .success();
```

Comment:
need to add this propagation delay before returning success: https://code.amazon.com/packages/AWS21Workflows/blobs/mainline/--/src/com/amazon/aws21/workflows/resources/sns/topicPolicy/TopicPolicyUpdateWorkflowImpl.java#:~:text=waitForCreate%20%3D%20createHandler.create(getResourceType()%2C%20topicArn%2C%20getStackId()%2C%20getLogicalResourceId()%2C%20waitForCreate)%3B